### PR TITLE
[Python] Expose DataFragment via dataset

### LINKED ISSUE #557: Expose DataFragment via Python SDK
# Problem Statement

By exposing the layout of the dataset, it could allow several distributed computing/training frameworks to load the Lance dataset in parallel as a stop-gap measure. 

For example, an envisioned workaround in PySpark to load Lance distributedly (without building the Spark DataSource first) could be  

```python

import lance
import pyspark

fragments = lance.dataset("uri").fragments()
fragments_df = pyspark.createDataFrame({"fragments": fragments}).repartition(num_of_executors)

fragmetns_df.rdd.flatmap(read_fragment).toDF(schema)
```

And a PyTorch distributed dataset could similar to our old PyTorch reader.

https://github.com/eto-ai/lance/blob/branch-0.2/python/lance/pytorch/data.py#L155-L164

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #557

### CODE REVIEW

File: python/python/lance/fragment.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+from __future__ import annotations
+from typing import Union, Optional, Iterator
+
+import pyarrow as pa
+
+
+class LanceFragment(pa.dataset.Fragment):
+    def __init__(self, dataset: "LanceDataset", fragment: LanceFragment):
+        self._ds = dataset
+        self._fragmnet = fragment
```
Comment (changhiskhan) [inline python/python/lance/fragment.py:27]: _fragment

File: python/python/lance/fragment.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    @property
+    def fragment_id(self):
+        return self._fragmnet.idw
```
Comment (changhiskhan) [inline python/python/lance/fragment.py:31]: What's "idw"?

File: python/python/lance/fragment.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    def count_rows(
+        self, filter: Optional[Union[pa.compute.Expression, str]] = None
+    ) -> int:
+        if filter is not None:
+            raise ValueError("Does not support filter at the moment")
+        return self._fragmnet.count_rows()
+
+    def head(self, num_rows: int) -> pa.Table:
+        batches = self.scanner(limit=num_rows).to_batches()
+        return pa.Table.from_batches(batches)
```
Comment (changhiskhan) [inline python/python/lance/fragment.py:42]: Why not call scanner.to_table directly?

File: python/python/lance/fragment.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        s = self._fragmnet.scanner(
+            columns=columns, filter=filter_str, limit=limit, offset=offset
+        )
+
+        from .dataset import LanceScanner
+
+        return LanceScanner(s, self._ds)
+
+    def take(self, indices, columns: Optional[list[str]] = None) -> pa.Table:
+        return pa.Table.from_batches([self._fragmnet.take(indices, columns=columns)])
```
Comment (changhiskhan) [inline python/python/lance/fragment.py:62]: Same here - does Take not return a Table?

File: rust/src/datatypes.rs [hunk 5]
```diff
@@ -692,6 +692,8 @@ impl Schema {
                     candidates.push(projected_field)
                 }
             } else {
+                use std::backtrace::Backtrace;
+                println!("Custom backtrace: {}", Backtrace::force_capture());
```
Comment (changhiskhan) [inline rust/src/datatypes.rs:696]: Debug prints ?

File: python/python/lance/fragment.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+"""Dataset Fragment"""
+
+from __future__ import annotations
+from typing import Union, Optional, Iterator
+
+import pyarrow as pa
+
+
+class LanceFragment(pa.dataset.Fragment):
+    def __init__(self, dataset: "LanceDataset", fragment: LanceFragment):
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/python/lance/fragment.py:25]: oh, this meant to be the pyo3 Fragment class.

Fixed.

File: python/python/lance/fragment.py (hunk 4 above)
Comment (eddyxu) [reply python/python/lance/fragment.py:62]: Yes this is same as Dataset.rs, that `take` returns batch in rust but convert to Table in python.

File: rust/src/datatypes.rs (hunk 5 above)
Comment (eddyxu) [reply rust/src/datatypes.rs:696]: k will remove this.

File: python/python/lance/fragment.py (hunk 3 above)
Comment (eddyxu) [reply python/python/lance/fragment.py:50]: Good pooint.

File: python/python/lance/fragment.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/fragment.py]: Typo. fixed.

File: python/python/lance/fragment.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/fragment.py]: fixed.

File: python/python/lance/fragment.py (hunk 4 above)
Comment (eddyxu) [reply python/python/lance/fragment.py]: i am fine either way. (i.e,. change both dataset / fragment take to return Table)

File: python/python/lance/dataset.py [hunk 7]
```diff
@@ -192,11 +193,15 @@ def replace_schema(self, schema: Schema):
         """
         raise NotImplementedError("not changing schemas yet")
 
-    def get_fragments(self, filter: Expression = None):
+    def get_fragments(self, filter: Expression = None) -> Iterator[pa.dataset.Fragment]:
         """
         Not implemented (just override pyarrow dataset to prevent segfault)
```
Comment (changhiskhan) [inline python/python/lance/dataset.py:198]: update the docstring?

File: python/python/lance/fragment.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    def count_rows(
+        self, filter: Optional[Union[pa.compute.Expression, str]] = None
+    ) -> int:
+        if filter is not None:
+            raise ValueError("Does not support filter at the moment")
+        return self._fragment.count_rows()
+
+    def head(self, num_rows: int) -> pa.Table:
+        batches = self.scanner(limit=num_rows).to_batches()
+        return pa.Table.from_batches(batches)
```
Comment (changhiskhan) [inline python/python/lance/fragment.py:50]: call self.scanner().to_table()?

**Review by changhiskhan — APPROVED** [review]
pending CI those last two nits

File: python/python/lance/fragment.py (hunk 8 above)
Comment (eddyxu) [reply python/python/lance/fragment.py]: fixed.

File: python/python/lance/dataset.py (hunk 7 above)
Comment (eddyxu) [reply python/python/lance/dataset.py]: fixed.

**Pull request merged**